### PR TITLE
HackStudio: Implement ToDoEntriesWidget

### PR DESCRIPTION
**Pull request by guerinoni**
Now it's possible to see all FIXME and TODO of the project.

Signed-off-by: Federico Guerinoni <guerinoni.federico@gmail.com>

### CODE REVIEW

**Comment by linusg** [comment]
Please remove the `Signed-off-by` line from the commit message.

**Comment by FalseHonesty** [comment]
Hmm, it feels to me like doing a simple text search for "TODO"/"FIXME" isn't the best way to approach this, because they should only be treated as actual todos when they are in a comment. Maybe a file's SyntaxHighlighter can communicate a list of TODOs it finds during lexing?

**Comment by guerinoni** [comment]
If reading these keywords from lexing is cheaper in terms of resources and time I can re-work in this way. If the effort is the same maybe I can just improve this search to be sure to find them only within a comment. :)

**Comment by FalseHonesty** [comment]
> If reading these keywords from lexing is cheaper in terms of resources and time I can re-work in this way. If the effort is the same maybe I can just improve this search to be sure to find them only within a comment. :)

Well, trying to find them in a comment is sort of just re-doing some of the work of the lexer, so might as well bundle the two together? Additionally, comment syntax is different for different languages, so it probably makes sense for the work to be done by something language aware, like the syntax highlighter, which has a different implementation depending on language.

**Comment by itamar8910** [comment]
Yeah, I agree with @FalseHonesty in that we should use a lexer's output to find TODOs, and i think it would make a lot of sense to use the language server for this.
It would also be less resource consuming because the language server lexes the code anyways.

We can have the language server look for TODOs within comment tokens and notify the client via an IPC message whenever it finds a TODO entry. This would also make the `ToDo Entries` view update live as you type :)
Essentially this can be quite similar to how the `ProjectDeclarations` gets updated from the language server and displayed in the `Locator`.

**Comment by guerinoni** [comment]
@FalseHonesty @itamar8910 
Thanks! Then I can add "TODO" in lexer cpp for now and then communicate all todo entries over IPC adding a special `handle`...
I'll try to work on it :)

**Comment by guerinoni** [comment]
I'm working on it :)

**Comment by itamar8910** [comment]
Very cool, I'll review this tomorrow :)

File: Userland/DevTools/HackStudio/ToDoEntries.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    void set_entries(const String& filename, const Vector<String>&& entries);
+
+    Vector<ToDoEntryPair> get_entries();
+
+    Function<void()> on_update = nullptr;
+
+private:
+    ToDoEntries() = default;
+    HashMap<String, Vector<String>> m_document_to_entris;
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/ToDoEntries.h:37]: typo: `entries`

File: Userland/DevTools/HackStudio/ToDoEntries.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+#pragma once
+
+#include <AK/Function.h>
+#include <AK/HashMap.h>
+#include <AK/Noncopyable.h>
+#include <AK/String.h>
+#include <LibGUI/AutocompleteProvider.h>
+#include <LibGUI/Icon.h>
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/ToDoEntries.h:14]: Are these two imports needed?

File: Userland/DevTools/HackStudio/LanguageServers/Shell/ClientConnection.h [hunk 3]
```diff
@@ -21,6 +21,9 @@ class ClientConnection final : public LanguageServers::ClientConnection {
         m_autocomplete_engine->set_declarations_of_document_callback = [this](const String& filename, Vector<GUI::AutocompleteProvider::Declaration>&& declarations) {
             async_declarations_in_document(filename, move(declarations));
         };
+        m_autocomplete_engine->set_todo_entries_of_document_callback = [this](const String& filename, Vector<String>&& todo_entris) {
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/LanguageServers/Shell/ClientConnection.h:24]: typo, `todo_entries` :)

File: Userland/DevTools/HackStudio/LanguageServers/Cpp/CppComprehensionEngine.cpp [hunk 4]
```diff
@@ -516,6 +516,12 @@ void CppComprehensionEngine::update_declared_symbols(DocumentData& document)
     set_declarations_of_document(document.filename(), move(declarations));
 }
 
+void CppComprehensionEngine::update_todo_entries(DocumentData& document)
+{
+    auto todo_entries = document.parser().get_todo_entries();
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/LanguageServers/Cpp/CppComprehensionEngine.cpp:521]: nitpick: no need for the `todo_entries` variable here

File: Userland/DevTools/HackStudio/LanguageServers/Cpp/ClientConnection.h [hunk 5]
```diff
@@ -22,6 +22,9 @@ class ClientConnection final : public LanguageServers::ClientConnection {
         m_autocomplete_engine->set_declarations_of_document_callback = [this](const String& filename, Vector<GUI::AutocompleteProvider::Declaration>&& declarations) {
             async_declarations_in_document(filename, move(declarations));
         };
+        m_autocomplete_engine->set_todo_entries_of_document_callback = [this](const String& filename, Vector<String>&& todo_entris) {
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/LanguageServers/Cpp/ClientConnection.h:25]: `todo_entries`

File: Userland/DevTools/HackStudio/LanguageServers/ClientConnection.h [hunk 6]
```diff
@@ -35,6 +35,9 @@ class ClientConnection
     virtual void auto_complete_suggestions(GUI::AutocompleteProvider::ProjectLocation const&) override;
     virtual void find_declaration(GUI::AutocompleteProvider::ProjectLocation const&) override;
 
+    static void set_declarations_of_document_callback(ClientConnection&, const String&, Vector<GUI::AutocompleteProvider::Declaration>&&);
+    static void set_todo_entries_of_document_callback(ClientConnection&, const String&, Vector<String>&&);
+
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/LanguageServers/ClientConnection.h:40]: Are these functions needed here in `ClientConnection.h`? I don't think they're ever used :thinking:

**Review by itamar8910 — CHANGES_REQUESTED** [review]
Please see my comments :)
Also, some nitpicking re. commit messages:
1. Please try to avoid unnecessary abbreviations such as `func`. `function` would have fit the commit title just fine :)
2. For `b35a27e`, the convention seems to be to just use `LanguageServers:` as the 'topic'.

File: Userland/DevTools/HackStudio/HackStudioWidget.cpp [hunk 7]
```diff
@@ -848,6 +849,7 @@ void HackStudioWidget::create_action_tab(GUI::Widget& parent)
     };
 
     m_find_in_files_widget = m_action_tab_widget->add_tab<FindInFilesWidget>("Find in files");
+    m_todo_entries_widget = m_action_tab_widget->add_tab<ToDoEntriesWidget>("ToDo Entries");
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/HackStudioWidget.cpp:852]: Also, let's name this tab just "TODO"

File: Userland/DevTools/HackStudio/LanguageServers/Cpp/ClientConnection.h (hunk 5 above)
Comment (guerinoni) [reply Userland/DevTools/HackStudio/LanguageServers/Cpp/ClientConnection.h:25]: Done

File: Userland/DevTools/HackStudio/HackStudioWidget.cpp (hunk 7 above)
Comment (guerinoni) [reply Userland/DevTools/HackStudio/HackStudioWidget.cpp:852]: Done

File: Userland/DevTools/HackStudio/LanguageServers/Cpp/CppComprehensionEngine.cpp (hunk 4 above)
Comment (guerinoni) [reply Userland/DevTools/HackStudio/LanguageServers/Cpp/CppComprehensionEngine.cpp:521]: Done :D

File: Userland/DevTools/HackStudio/ToDoEntries.h (hunk 2 above)
Comment (guerinoni) [reply Userland/DevTools/HackStudio/ToDoEntries.h:14]: Good catch!

File: Userland/Libraries/LibCpp/Parser.h [hunk 8]
```diff
@@ -35,6 +35,7 @@ class Parser final {
     void print_tokens() const;
     const Vector<String>& errors() const { return m_state.errors; }
     const Preprocessor::Definitions& preprocessor_definitions() const { return m_preprocessor_definitions; }
+    Vector<String> get_todo_entries();
```
[the remark this answers is not in the document]
Comment (guerinoni) [reply Userland/Libraries/LibCpp/Parser.h:38]: Done!

File: Userland/DevTools/HackStudio/LanguageServers/ClientConnection.h (hunk 6 above)
Comment (guerinoni) [reply Userland/DevTools/HackStudio/LanguageServers/ClientConnection.h]: Solved :)

**Review by itamar8910 — APPROVED** [review]
LGTM, thanks for working on this! :tada:

**Review by linusg — CHANGES_REQUESTED** [review]
You have made various fixes suggested for either the first or second commit in the third commit. Please don't do that, squash them into the appropriate preceding commit instead.

The comments regarding commit messages have been ignored.

**Comment by guerinoni** [comment]
@linusg thank you for remind me :) now should be consistent

**Comment by itamar8910** [comment]
Please see my previous comments regarding commit messages:
> 1. Please try to avoid unnecessary abbreviations such as `func`. `function` would have fit the commit title just fine :)
> 2. For `b35a27e`, the convention seems to be to just use `LanguageServers:` as the 'topic'.

**Review by linusg — COMMENTED** [review]
Still:

> Please try to avoid unnecessary abbreviations such as func. function would have fit the commit title just fine :)

Specific suggestions to make this easier:

- `LibCpp: Add function for retrieving TODO comments from the parser`
- `LanguageServers: Add function to collect TODO entries in a document`

File: Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                return Gfx::FontDatabase::default_fixed_width_font();
+            return {};
+        }
+        if (role == GUI::ModelRole::Display) {
+            auto& match = m_matches.at(index.row());
+            switch (index.column()) {
+            case Column::Filename:
+                return match.filename;
+            case Column::Text:
+                return match.comment;
```
Comment (alimpfard) [inline Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp:63]: Not knowing the source position of a Todo/FIXME kinda just defeats the whole purpose of a widget like this
There would be lots of context-dependent to-dos without their context, and no way to get to it...

**Comment by alimpfard** [comment]
Also, please write the commit messages in proper English, the ones @linusg suggested could serve as examples (or you could just use them as-is)

File: Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp (hunk 9 above)
Comment (guerinoni) [reply Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp:63]: Maybe with line can be useful? Or click to bring the cursor in that position...

Comment (alimpfard) [reply Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp:63]: Yep.
I think LibCpp already knows about token positions, so you can probably just use those to calculate the positions for the TODOs/FIXMEs.

Comment (guerinoni) [reply Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp:63]: Now I pass a `Vector<String>` through the functions, maybe I need some struct of TODO entry to pass

Comment (guerinoni) [reply Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp]: Add position and click to bring the cursor on that TODO

File: Userland/Libraries/LibCpp/Preprocessor.h [hunk 10]
```diff
@@ -14,6 +14,14 @@
 #include <AK/Vector.h>
 
 namespace Cpp {
+
+struct TodoEntry {
```
Comment (itamar8910) [inline Userland/Libraries/LibCpp/Preprocessor.h:18]: Does the `TodoEntry` struct belong in `Preprocessor.h`?
I think a better place to define it in would be in `Parser.h`, just before the `get_todo_entries()` function.

File: Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+ToDoEntriesWidget::ToDoEntriesWidget()
+{
+    set_layout<GUI::VerticalBoxLayout>();
+    m_result_view = add<GUI::TableView>();
+
+    m_result_view->on_activation = [](auto& index) {
+        auto& match = *(const Cpp::TodoEntry*)index.internal_data();
+        open_file(match.filename);
+        current_editor().set_selection({ { match.line, match.column }, { match.line, match.column } });
+        current_editor().set_focus(true);
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp:110]: I think using the version of `open_file` that takes a line&column would be better here.
(Also, I don't think we really need to have the TODO line selected, just jumping to it in the source is enough).
```suggestion
        open_file(match.filename, match.line, match.column);
```

File: Userland/DevTools/HackStudio/AutoCompleteResponse.h [hunk 12]
```diff
@@ -102,4 +103,33 @@ inline bool decode(Decoder& decoder, GUI::AutocompleteProvider::Declaration& dec
     return true;
 }
 
+template<>
+inline bool encode(Encoder& encoder, const Cpp::TodoEntry& entry)
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/AutoCompleteResponse.h:107]: According to the new conventions, new code should use east consts (Please also correct this in other places in the code).

**Review by itamar8910 — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCpp/Preprocessor.h (hunk 10 above)
Comment (guerinoni) [reply Userland/Libraries/LibCpp/Preprocessor.h]: Done :)

File: Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp (hunk 11 above)
Comment (guerinoni) [reply Userland/DevTools/HackStudio/ToDoEntriesWidget.cpp]: Done :)

File: Userland/DevTools/HackStudio/AutoCompleteResponse.h (hunk 12 above)
Comment (guerinoni) [reply Userland/DevTools/HackStudio/AutoCompleteResponse.h]: Done :)

**Review by itamar8910 — APPROVED** [review]
Okay, looks good to me.

**Comment by guerinoni** [comment]
@linusg @alimpfard do you think this can be ok? :^)

**Comment by itamar8910** [comment]
Bumping this, @linusg @alimpfard :smile:

File: Userland/DevTools/HackStudio/ToDoEntries.cpp [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    if (on_update)
+        on_update();
+}
+
+Vector<Cpp::Parser::TodoEntry> ToDoEntries::get_entries()
+{
+    Vector<Cpp::Parser::TodoEntry> ret;
+    for (auto& it : m_document_to_entries)
+        for (auto& entry : it.value)
+            ret.append({ entry.content, it.key, entry.line, entry.column });
```
Comment (linusg) [inline Userland/DevTools/HackStudio/ToDoEntries.cpp:29]: Please add curly braces to any if/else/for/etc body with more than one line:

```suggestion
    for (auto& it : m_document_to_entries) {
        for (auto& entry : it.value)
            ret.append({ entry.content, it.key, entry.line, entry.column });
    }
```

**Review by linusg — COMMENTED** [review]
thanks @itamar8910 :smile:

one nitpick, but the 4th commit is a bit of a mess and does much more than it claims to - namely all the east-const conversions for code added in the previous commits. - those should be squashed instead.

File: Userland/DevTools/HackStudio/ToDoEntries.cpp (hunk 13 above)
Comment (guerinoni) [reply Userland/DevTools/HackStudio/ToDoEntries.cpp]: I can't squash them because I also changed some struct in function parameter

**Pull request merged**